### DISCUSSION TITLE
stumpy.motifs question regarding "cutoff" and "max_distance" parameter?

### DISCUSSION THREAD

**Discussion by jmrichardson** [TDAmeritrade/stumpy#548] [Q&A] [answered]
Hi, I was hoping someone could help me better understand the "cutoff" parameter in "stumpy.motifs".  From docs:

```
    cutoff : float, default None
        The largest matrix profile value (distance) that a candidate motif is allowed
        to have. If `None`, this defaults to
        `np.nanmax([np.nanmean(P) - 2.0 * np.nanstd(P), np.nanmin(P)])`
```

I do understand that max_distance is the maximum distance that a motif can have to it's neighbor.  But I don't understand how cutoff is used.  Thanks for any help.

**Comment by seanlaw** [comment, marked as answer]
@jmrichardson Thank you for your question and welcome to the STUMPY community. 

> I do understand that max_distance is the maximum distance that a motif can have to it's neighbor. But I don't understand how cutoff is used. Thanks for any help.

So, after you've computed the matrix profile, you end up with an array of distances (to their one-nearest-neighbor). Now, say you are interested in finding a motif AND a list of their top matches but, first, you need to scan along your matrix profile and pick a motif. How do you decide which subsequence is a "good" candidate? Typically, you'd choose the subsequence with the smallest matrix profile value as your top motif and then look for "matches" to this motif in your original time series. But after this first subsequence, how would you find your next "candidate motif"? Well, you'd repeat the same process and grab the next smallest matrix profile value and so on.

However, let's say the first candidate motif had a matrix profile value of `0.1` and the second candidate motif had a matrix profile value of `0.2`. Both of these values are pretty small and probably "comparable" to each other and they both have a nearest neighbor that is very similar to the motif. But what happens when the third candidate motif has a matrix profile value of `95.7`? This matrix profile value is much, much bigger than `0.1` or `0.2` and it tells you that the one-nearest-neighbor to the third motif is actually quite far away (compared with candidate motifs 1 and 2). And since we are traversing the candidates in sorted order, you'll be sure that all subsequent candidate motifs will only have matrix profile distances that are even larger than `95.7`. So, `cutoff` allows you to choose the largest matrix profile value that you'd allow for a subsequence to be considered as a candidate motif. By default, this value is set by taking your full matrix profile and computing the point that is two standard deviations below the mean matrix profile value (i.e., small matrix profile values that are in the bottom 2.5% compared with their peers). Of course, in this example, we might choose, say, a cutoff of `0.5` and ignore all candidates subsequences that have a matrix profile value greater than `0.5`. This means, that we'd prune away all candidates and this results in only returning 2 motifs. Of course, setting `cutoff = 95.0` would have the same result of only 2 motifs. This just helps to ensure that we don't scan all subsequences.

Does that make sense?

**Reply by jmrichardson** [reply]
Hi @seanlaw ,

Thank you for the fast and thorough response!  It does make perfect sense and adds a lot of flexibility to finding the motifs!  This library is amazing!  Thank you for providing it to the community :)

**Reply by seanlaw** [reply]
That's great news! Are you able to share a little bit of how and where you are leveraging STUMPY?

**Reply by jmrichardson** [reply]
Hi @seanlaw ,

I have been using ML for stock market price prediction and have been unimpressed with its capabilities especially with comparative feature pattern recognition.  ML does not capture information the way humans do in their ability to recognize time series patterns across multiple features.  DL theoretically has this capability but often requires much more data than for example stock history.  I have been looking for a way to recognize persistent patterns such as breakouts, flags, etc.  I am currently working on pattern recognition using stumpy and very impressed with how effective it is.  Currently, I am working on recognition as well as probability of matching, and feeding those signals into a larger framework that manages my ML workflow.  Although I haven't completed the project, it does look very promising via the plot visuals of motifs that have been found across a basket of equities.   I am currently doing the following for a multiple close price series which I obtained from the "Consensus Motif Search" tutorial:

- Append a np.nan to the end of each time series. This is used to identify the boundary between neighboring time series and ensures that any identified motif will not straddle multiple time series.
- Concatenate all of the time series into a single long time series
- Compute the matrix profile (self-join) on the aforementioned concatenated time series

Next, I am using the stumpy.motifs function to find candidate motifs.  The returned "motif_distances" is very helpful as you can use it to determine the number of matches for a specific motif.  From these motifs, I am using a trend scanner to find an appropriate cut point for which I will replace the data after the cut point in each motif with np.nan and then use stumpy.match to find the corresponding motif in the original time series.  I am working on this now, but after finding these "cut" motifs, determine the probability of all that were found to behave the same as the orginal motif.  At least that is the plan and hope that it makes sense.  Ultimately, I will be feeding these binary signals as features to my ML framework.

Question if you don't mind me asking, it appears that TDA sponsors this project(?).  Are you or TDA using stumpy in the financial markets?

**Reply by fritshermans** [reply]
@seanlaw, could you please explain how `max_distance` plays a role in your example above?

**Reply by seanlaw** [reply]
@fritshermans As stated in the [stumpy.motifs](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.motifs) documentaion:

> max_distance – For a candidate motif, Q, and a non-trivial subsequence, S, max_distance is the maximum distance allowed between Q and S so that S is considered a match of Q. If max_distance is a function, then it must be a function that accepts a single parameter, D, in its function signature, which is the distance profile between Q and T. If None, this defaults to np.nanmax([np.nanmean(D) - 2.0 * np.nanstd(D), np.nanmin(D)]).

So, starting with the pre-computed matrix profile, let's say you have a subsequence, Q, that has a distance to its one-nearest-neighbor that is below `cutoff`, then this subsequence is considered a "candidate motif". What we then do is compute the distance profile (i.e., compute the pairwise distances between `Q` and EVERY non-trivial subsequence along the time series) and we extract the 1st best (non-trivial) match and this subsequence, `S1`, should be the same as the matrix profile distance. However, if this distance, `D1`, is larger than `max_distance` then we disallow it to be returned as a "match". If `D1` is smaller than `max_distance` then we allow it to be returned as a match. For illustration purposes, imagine that `D1` was indeed smaller than `max_distance` and so we will return `S1` as a match. Next, we search for the second-nearest neighbor for `Q` and we find subsequence `S2`. However, `S2` has a distance that is larger than `max_distance` from `Q` and so we stop searching altogether and we only return `S1` as the only possible match according to the `max_distance` criteria. 

@fritshermans Please let me know if that makes sense

**Reply by fritshermans** [reply]
Now it's clear to me. Thanks a lot!
